### ISSUE TITLE
Autolevel adjust incorreclty

### ISSUE THREAD

**Issue by Sungod3000** [MarlinFirmware/Marlin#2665]
Hi,

I tested this version and while trying to find the best auto level routine. The g30 and g29 work reliably and as expected. When I visualize the result of g29 I get a repeatable pattern. 

I printed some smaller items over the last days and it worked well however im noticing now that I print bigger item that the plane is quite tilted. I checked the marlin wiki to find out more, but it only lists the g29 parameters.

I didn't find a way to check what leveling compensation matrix is applied or is there a parameter that applies the matrix that is not included in the default g29?

**Comment by AnHardt** [comment]
" this version" is not enough information. MarlinFirmware/Marlin has today 4 branches and a couple of .zips you could have downloaded.
If you are on one of the newer branches a "G29 V4" may tell you more about the matrix and the height map. Remember not to G28 after a G29 - else the matrix is reset. If the height-map appears not only tilted but saddled ABL can not do much against that. 

A link to your 'Configuration.h' and a printout of "G29 V4", or at least "G29" will help us to analyse your case.

**Comment by Sungod3000** [comment]
my bad I thought each branch has its own issues section. im using 1.1 RC1
-g29 is only tilted http://www.maui-3d.com/cgi-bin/plotG29?plotG29ID=5370
-no g28 after g29

```
SENDING:G29 V4
G29 Auto Bed Leveling
Bed X: 80.000 Y: -84.000 Z: 1.719
Bed X: 39.000 Y: -84.000 Z: 1.200
Bed X: -2.000 Y: -84.000 Z: 0.981
Bed X: -43.000 Y: -84.000 Z: 0.725
Bed X: -84.000 Y: -84.000 Z: 0.450
Bed X: -84.000 Y: -43.000 Z: 0.438
Bed X: -43.000 Y: -43.000 Z: 0.787
Bed X: -2.000 Y: -43.000 Z: 1.012
Bed X: 39.000 Y: -43.000 Z: 1.275
Bed X: 80.000 Y: -43.000 Z: 1.556
Bed X: 121.000 Y: -2.000 Z: 1.700
Bed X: 80.000 Y: -2.000 Z: 1.175
Bed X: 39.000 Y: -2.000 Z: 1.000
Bed X: -2.000 Y: -2.000 Z: 0.794
Bed X: -43.000 Y: -2.000 Z: 0.575
Bed X: -84.000 Y: -2.000 Z: 0.325
Bed X: -84.000 Y: 39.000 Z: 0.431
Bed X: -43.000 Y: 39.000 Z: 0.644
Bed X: -2.000 Y: 39.000 Z: 0.912
Bed X: 39.000 Y: 39.000 Z: 1.056
Bed X: 80.000 Y: 39.000 Z: 1.212
Bed X: 80.000 Y: 80.000 Z: 1.106
Bed X: 39.000 Y: 80.000 Z: 0.900
Bed X: -2.000 Y: 80.000 Z: 0.762
Bed X: -43.000 Y: 80.000 Z: 0.575
Bed X: -84.000 Y: 80.000 Z: 0.337
Bed X: -2.000 Y: 121.000 Z: 0.969
-0.87 -0.64 -0.42 -0.13 0.04 0.31 1.08
-0.91 -0.63 -0.36 -0.10 0.12 0.64 1.03
-0.99 -0.64 -0.29 -0.07 0.19 0.48 0.76
-1.01 -0.76 -0.51 -0.29 -0.08 0.09 0.62
-0.86 -0.65 -0.44 -0.17 -0.02 0.13 0.29
-0.79 -0.74 -0.51 -0.32 -0.18 0.03 0.23
-1.05 -0.84 -0.57 -0.11 -0.34 -0.19 0.08
```

**Comment by KiteLab** [comment]
Ether this is not the printout of a "G29 V4" from 1.1 RC1, or you edited almost all information away.
It looks more like from a much older branch.

**Comment by KiteLab** [comment]
Ah - sorry. You have a delta? There is no matrix for deltas. ABL works totally different there.

**Comment by Sungod3000** [comment]
Yes delta.

**Comment by Roxy-3D** [comment]
SunGod3000 Do you have your bed mounted with 3 Screws around the perimeter?   If so, I may have a piece of code that will help you manually level your bed so you don't need the Auto Bed Leveling.

**Comment by Sungod3000** [comment]
I have levelled it manually and it works (never the less Im interested in your code). So ABL is not necessary but its really convenient.

**Comment by Roxy-3D** [comment]
OK, Understood.   I'm bringing my Delta printer up but it isn't fully functional yet.  So I can't actually dig into the G29 yet.   But to make sure the bed was physically level, I wrote a small piece of code to probe (and measure) very close to the screws for the bed.   

Add this before the G28 code:

```
    case 27:    // Roxy Probe 3 points to manually level bed
      {
      ProbeAction p1, p2, p3;
        p1 = ProbeDeploy; 
    p2 = ProbeStay; 
    p3 = ProbeStow;

      // Probe at 3 arbitrary points
      float z_at_pt_1, z_at_pt_2, z_at_pt_3;

    SERIAL_PROTOCOLPGM("Outer Circle of Points:\n");

    z_at_pt_1 = probe_pt(ABL_PROBE_PT_1_X, ABL_PROBE_PT_1_Y, Z_RAISE_BEFORE_PROBING, p1, 4),
    z_at_pt_2 = probe_pt(ABL_PROBE_PT_2_X, ABL_PROBE_PT_2_Y, current_position[Z_AXIS] + Z_RAISE_BETWEEN_PROBINGS, p2, 4),
    z_at_pt_3 = probe_pt(ABL_PROBE_PT_3_X, ABL_PROBE_PT_3_Y, current_position[Z_AXIS] + Z_RAISE_BETWEEN_PROBINGS, p2, 4);

    SERIAL_PROTOCOLPGM("Middle Circle of Points:\n");

    z_at_pt_1 = probe_pt(ABL_PROBE_PT_1_X/1.75, ABL_PROBE_PT_1_Y/1.75, Z_RAISE_BEFORE_PROBING, p2, 4),
    z_at_pt_2 = probe_pt(ABL_PROBE_PT_2_X/1.75, ABL_PROBE_PT_2_Y/1.75, current_position[Z_AXIS] + Z_RAISE_BETWEEN_PROBINGS, p2, 4),
    z_at_pt_3 = probe_pt(ABL_PROBE_PT_3_X/1.75, ABL_PROBE_PT_3_Y/1.75, current_position[Z_AXIS] + Z_RAISE_BETWEEN_PROBINGS, p2, 4);

    SERIAL_PROTOCOLPGM("Inner Circle of Points:\n");

    z_at_pt_1 = probe_pt(ABL_PROBE_PT_1_X/3.0, ABL_PROBE_PT_1_Y/3.0, Z_RAISE_BEFORE_PROBING, p2, 4),
    z_at_pt_2 = probe_pt(ABL_PROBE_PT_2_X/3.0, ABL_PROBE_PT_2_Y/3.0, current_position[Z_AXIS] + Z_RAISE_BETWEEN_PROBINGS, p2, 4),
    z_at_pt_3 = probe_pt(ABL_PROBE_PT_3_X/3.0, ABL_PROBE_PT_3_Y/3.0, current_position[Z_AXIS] + Z_RAISE_BETWEEN_PROBINGS, p3, 4);

      clean_up_after_endstop_move();
//      set_bed_level_equation_3pts(z_at_pt_1, z_at_pt_2, z_at_pt_3);
    }
    break;
```

and you may need to add this block of code to get the U shaped dip from happening:

```
#define MAX_MOVEMENT_WITHOUT_BREAKING_UP_MOVE   25
  static void do_blocking_move_to(float x, float y, float z) {
    float oldFeedRate = feedrate;
    #if ENABLED(DEBUG_LEVELING_FEATURE)
      if (marlin_debug_flags & DEBUG_LEVELING) {
        print_xyz("do_blocking_move_to", x, y, z);
      }
    #endif
    #if ENABLED(DELTA)
      feedrate = XY_TRAVEL_SPEED;
 //
 // Should we break up the move because it is too big ???
 //
      if (abs(x-current_position[X_AXIS]) > MAX_MOVEMENT_WITHOUT_BREAKING_UP_MOVE   ||
          abs(y-current_position[Y_AXIS]) > MAX_MOVEMENT_WITHOUT_BREAKING_UP_MOVE   ||
          abs(z-current_position[Z_AXIS]) > MAX_MOVEMENT_WITHOUT_BREAKING_UP_MOVE ) {

         do_blocking_move_to( current_position[X_AXIS] + (x-current_position[X_AXIS])/2.0,
                              current_position[Y_AXIS] + (y-current_position[Y_AXIS])/2.0,
                              current_position[Z_AXIS] + (z-current_position[Z_AXIS])/2.0);
//
// At this point, we are halfway to where we want to be.  Do the other half of the move.
//
        do_blocking_move_to(x, y, z);
        feedrate = oldFeedRate;
        return;
      } else {      // No...  It is small enough to do in one piece.
        destination[X_AXIS] = x;
        destination[Y_AXIS] = y;
        destination[Z_AXIS] = z;
        prepare_move_raw(); // this will also set_current_to_destination
        st_synchronize();
      }
    #else
      feedrate = homing_feedrate[Z_AXIS];
      current_position[Z_AXIS] = z;
      line_to_current_position();
      st_synchronize();
      feedrate = xy_travel_speed;
      current_position[X_AXIS] = x;
      current_position[Y_AXIS] = y;
      line_to_current_position();
      st_synchronize();
    #endif
    feedrate = oldFeedRate;
  }
```

And in Configuration.h set the 3-Point leveling points to points very close to your set screws.  And then ~~move them such that they are always turned on if you have Auto Bed Leveling turned on.~~ just delete the #else immediately above them.   More features coming later today on this.

**Comment by sarf2k4** [comment]
Hi, I think i had same problem before, the bed was tilted on the right side marlin registered as lower, left as higher, in the end during prints, the nozzle were scraping the bed on the right side. I decided to try cleaning my glass bed, put a new 3-4 layer of glue stick and test again, my problem somewhat gone.

Just my suggestion that might help

**Comment by Sungod3000** [comment]
@sarf2k4 yes thats the pattern. Im using aluminium and I cleaned all the hairspray of at one time, but it didnt change anything.

**Comment by sarf2k4** [comment]
I had the same last time when I switched to aluminium with inductive sensor, in the end I removed it because the nozzle isn't near enough. Not to mention that there's no way to see the 'current' z offset without lcd panel

The pattern I got before cleaning the glass were always descend at right side, ascend at left side, no matter how I tried to adjust the bed (once only) and limited to small prints only mostly. 

What about lowering your homing speed? I use around 45*4 I think. I saw in another issue topics saying lowering the speed would have higher accuracies

**Comment by Sungod3000** [comment]
my homing speed is low already and the repeatability of the pattern is very good, so i dont doubt the accuracy of the sensor. the z offset show in the m502 read out btw.

i will try the old RichCattel version soon, and see if that gives me a different result

**Comment by sarf2k4** [comment]
I ran into same problem last night, it really does adjusts incorrectly; having the nozzle raised on the left, lowered on the right. Now that I suspect there's something wrong with the mesh bed leveling being flipped making marlin produces this problem. 

For example in the physical world, the left side of the bed is lower and right side is higher, after the g29 routine, the marlin would register that on the left side is higher while right side is lower, thus during printing the nozzle are being pushed to scrape the bed surface at right side

This is just an assumptions, I attached an image showing the printer is printing on the first layer, originally the nozzle were too far; I used babystepping to lower the z axis, but the first layer were inconsistent, the right side of the object, I raised the z axis a little bit since my extruder motor are skipping steps because the nozzle are somewhat scraping the bed

http://imgur.com/a/406Kf

After the print on the image above, I tried to physically raise the right side of the bed, probably by around 0.2-0.5, screw turned by about quarter counter clockwise. The results were the layer are somewhat uniform. 

Although I don't understand why, the print before the object in the image were also quite consistent too

I'm joining in this discussion to give my suggestions and opinion regarding this. Auto Bed Leveling aren't supposed to make a lot of problems

**Comment by thinkyhead** [comment]
@Sungod3000 @sarf2k4 Are you still seeing this issue with the latest RC? If so, be sure to enable `DEBUG_LEVELING_FEATURE` and use `M115 S32` to turn on debugging for homing and leveling. Then you will get a detailed log when you do `G29`. That should provide a clue where it's messing up.

**Comment by Sungod3000** [comment]
I disabled auto level but I remember that someone on the reprap forum suggested some verbose commands but they didnt yield any new information. Im making a new effector that will have space for my leveling probe, Ill give a new shot then.

**Comment by sarf2k4** [comment]
Sorry I didn't notice you were replying to us, I probably solved it on my own printer, turns out to be adjusting the left and right motor did the trick (prusa i3 based) and make it leveled. But I still having same problem on my friend's printer, did the same but problem persist as well, I'm providing you the config.h, and verbose output for the bed at 0'c (idle),70'c and 110'c. The problem only occurred during at 70'c before; printing pla on bare glass without any gluestick. As of now after I insulated the heat to reach 110'c to print abs, it is good as far as I know (didn't do big prints yet on abs) and haven't been printing any pla ever since 2 days ago

Here is some [extra details](http://forums.reprap.org/read.php?262,629751) that you might need. Note that the one in the link is before I insulated the heatbed with cardboard

Bed at Idle:

```
SENDING:G29 V4
G29 Auto Bed Leveling
Bed X: 15.000 Y: 20.000 Z: 2.466
Bed X: 100.000 Y: 20.000 Z: 2.837
Bed X: 185.000 Y: 20.000 Z: 3.015
Bed X: 185.000 Y: 97.000 Z: 2.846
Bed X: 100.000 Y: 97.000 Z: 2.711
Bed X: 15.000 Y: 97.000 Z: 2.102
Bed X: 15.000 Y: 174.000 Z: 2.089
Bed X: 100.000 Y: 174.000 Z: 2.256
Bed X: 185.000 Y: 174.000 Z: 2.840
Eqn coefficients: a: 0.00400850 b: -0.00245239 d: 2.41051340
Mean of sampled points: 2.57348108
planeNormal x: -0.004008 y: 0.002452 z: 1.000000

Bed Height Topography:
+-----------+
|...Back....|
|Left..Right|
|...Front...|
+-----------+
 -0.48481 -0.31732 +0.26642
 -0.47126 +0.13711 +0.27258
 -0.10796 +0.26396 +0.44130

Corrected Bed Height vs. Bed Topology:
 +0.17528 +0.00205 +0.24508
 +0.00000 +0.26765 +0.06240
 +0.17447 +0.20567 +0.04229

Bed Level Correction Matrix:
+0.999992 +0.000000 +0.004008
+0.000010 +0.999997 -0.002452
-0.004008 +0.002452 +0.999989
```

Bed at 70'c

```
SENDING:G29 V4
G29 Auto Bed Leveling
Bed X: 15.000 Y: 20.000 Z: 2.409
Bed X: 100.000 Y: 20.000 Z: 2.830
Bed X: 185.000 Y: 20.000 Z: 3.065
Bed X: 185.000 Y: 97.000 Z: 2.860
Bed X: 100.000 Y: 97.000 Z: 2.663
Bed X: 15.000 Y: 97.000 Z: 2.078
Bed X: 15.000 Y: 174.000 Z: 2.060
Bed X: 100.000 Y: 174.000 Z: 2.256
Bed X: 185.000 Y: 174.000 Z: 2.889
Eqn coefficients: a: 0.00444557 b: -0.00237775 d: 2.35381889
Mean of sampled points: 2.56773400
planeNormal x: -0.004446 y: 0.002378 z: 1.000000

Bed Height Topography:
+-----------+
|...Back....|
|Left..Right|
|...Front...|
+-----------+
 -0.50739 -0.31158 +0.32143
 -0.49015 +0.09483 +0.29187
 -0.15887 +0.26232 +0.49754

Corrected Bed Height vs. Bed Topology:
 +0.18206 +0.00000 +0.25514
 +0.01622 +0.22332 +0.04250
 +0.16441 +0.20773 +0.06508

Bed Level Correction Matrix:
+0.999990 +0.000000 +0.004446
+0.000011 +0.999997 -0.002378
-0.004446 +0.002378 +0.999987
```

Bed at 110'c

```
SENDING:G29 V4
G29 Auto Bed Leveling
Bed X: 15.000 Y: 20.000 Z: 2.499
Bed X: 100.000 Y: 20.000 Z: 2.914
Bed X: 185.000 Y: 20.000 Z: 3.094
Bed X: 185.000 Y: 97.000 Z: 2.876
Bed X: 100.000 Y: 97.000 Z: 2.728
Bed X: 15.000 Y: 97.000 Z: 2.095
Bed X: 15.000 Y: 174.000 Z: 2.091
Bed X: 100.000 Y: 174.000 Z: 2.333
Bed X: 185.000 Y: 174.000 Z: 2.898
Eqn coefficients: a: 0.00427895 b: -0.00256435 d: 2.43483209
Mean of sampled points: 2.61398458
planeNormal x: -0.004279 y: 0.002564 z: 1.000000

Bed Height Topography:
+-----------+
|...Back....|
|Left..Right|
|...Front...|
+-----------+
 -0.52285 -0.28147 +0.28380
 -0.51916 +0.11385 +0.26163
 -0.11522 +0.29981 +0.47961

Corrected Bed Height vs. Bed Topology:
 +0.19376 +0.07143 +0.27299
 +0.00000 +0.26930 +0.05337
 +0.20649 +0.25781 +0.07390

Bed Level Correction Matrix:
+0.999991 +0.000000 +0.004279
+0.000011 +0.999997 -0.002564
-0.004279 +0.002564 +0.999988
```

``` cpp
#ifndef CONFIGURATION_H
#define CONFIGURATION_H

#include "boards.h"
#include "macros.h"

//===========================================================================
//============================= Getting Started =============================
//===========================================================================
/*
Here are some standard links for getting your machine calibrated:
 * http://reprap.org/wiki/Calibration
 * http://youtu.be/wAL9d7FgInk
 * http://youtu.be/wAL9d7FgInk
 * http://calculator.josefprusa.cz
 * http://reprap.org/wiki/Triffid_Hunter%27s_Calibration_Guide
 * http://www.thingiverse.com/thing:5573
 * https://sites.google.com/site/repraplogphase/calibration-of-your-reprap
 * http://www.thingiverse.com/thing:298812
*/

// This configuration file contains the basic settings.
// Advanced settings can be found in Configuration_adv.h
// BASIC SETTINGS: select your board type, temperature sensor type, axis scaling, and endstop configuration

//===========================================================================
//============================= DELTA Printer ===============================
//===========================================================================
// For a Delta printer replace the configuration files with the files in the
// example_configurations/delta directory.
//

//===========================================================================
//============================= SCARA Printer ===============================
//===========================================================================
// For a Scara printer replace the configuration files with the files in the
// example_configurations/SCARA directory.
//

// @section info

#if ENABLED(USE_AUTOMATIC_VERSIONING)
  #include "_Version.h"
#else
  #include "Default_Version.h"
#endif

// User-specified version info of this build to display in [Pronterface, etc] terminal window during
// startup. Implementation of an idea by Prof Braino to inform user that any changes made to this
// build by the user have been successfully uploaded into firmware.
#define STRING_CONFIG_H_AUTHOR "(none, default config)" // Who made the changes.
#define SHOW_BOOTSCREEN
#define STRING_SPLASH_LINE1 SHORT_BUILD_VERSION // will be shown during bootup in line 1
//#define STRING_SPLASH_LINE2 STRING_DISTRIBUTION_DATE // will be shown during bootup in line 2

// @section machine

// SERIAL_PORT selects which serial port should be used for communication with the host.
// This allows the connection of wireless adapters (for instance) to non-default port pins.
// Serial port 0 is still used by the Arduino bootloader regardless of this setting.
// :[0,1,2,3,4,5,6,7]
#define SERIAL_PORT 0

// This determines the communication speed of the printer
// :[2400,9600,19200,38400,57600,115200,250000]
#define BAUDRATE 250000


// Enable the Bluetooth serial interface on AT90USB devices
//#define BLUETOOTH

// The following define selects which electronics board you have.
// Please choose the name from boards.h that matches your setup
#ifndef MOTHERBOARD
  #define MOTHERBOARD BOARD_SANGUINOLOLU_12
#endif

// Optional custom name for your RepStrap or other custom machine
// Displayed in the LCD "Ready" message
//#define CUSTOM_MACHINE_NAME "3D Printer"

// Define this to set a unique identifier for this printer, (Used by some programs to differentiate between machines)
// You can use an online service to generate a random UUID. (eg http://www.uuidgenerator.net/version4)
//#define MACHINE_UUID "00000000-0000-0000-0000-000000000000"

// This defines the number of extruders
// :[1,2,3,4]
#define EXTRUDERS 1

// Offset of the extruders (uncomment if using more than one and relying on firmware to position when changing).
// The offset has to be X=0, Y=0 for the extruder 0 hotend (default extruder).
// For the other hotends it is their distance from the extruder 0 hotend.
//#define EXTRUDER_OFFSET_X {0.0, 20.00} // (in mm) for each extruder, offset of the hotend on the X axis
//#define EXTRUDER_OFFSET_Y {0.0, 5.00}  // (in mm) for each extruder, offset of the hotend on the Y axis


//// The following define selects which power supply you have. Please choose the one that matches your setup
// 1 = ATX
// 2 = X-Box 360 203Watts (the blue wire connected to PS_ON and the red wire to VCC)
// :{1:'ATX',2:'X-Box 360'}

#define POWER_SUPPLY 1

// Define this to have the electronics keep the power supply off on startup. If you don't know what this is leave it.
//#define PS_DEFAULT_OFF

// @section temperature

//===========================================================================
//============================= Thermal Settings ============================
//===========================================================================
//
//--NORMAL IS 4.7kohm PULLUP!-- 1kohm pullup can be used on hotend sensor, using correct resistor and table
//
//// Temperature sensor settings:
// -2 is thermocouple with MAX6675 (only for sensor 0)
// -1 is thermocouple with AD595
// 0 is not used
// 1 is 100k thermistor - best choice for EPCOS 100k (4.7k pullup)
// 2 is 200k thermistor - ATC Semitec 204GT-2 (4.7k pullup)
// 3 is Mendel-parts thermistor (4.7k pullup)
// 4 is 10k thermistor !! do not use it for a hotend. It gives bad resolution at high temp. !!
// 5 is 100K thermistor - ATC Semitec 104GT-2 (Used in ParCan & J-Head) (4.7k pullup)
// 6 is 100k EPCOS - Not as accurate as table 1 (created using a fluke thermocouple) (4.7k pullup)
// 7 is 100k Honeywell thermistor 135-104LAG-J01 (4.7k pullup)
// 71 is 100k Honeywell thermistor 135-104LAF-J01 (4.7k pullup)
// 8 is 100k 0603 SMD Vishay NTCS0603E3104FXT (4.7k pullup)
// 9 is 100k GE Sensing AL03006-58.2K-97-G1 (4.7k pullup)
// 10 is 100k RS thermistor 198-961 (4.7k pullup)
// 11 is 100k beta 3950 1% thermistor (4.7k pullup)
// 12 is 100k 0603 SMD Vishay NTCS0603E3104FXT (4.7k pullup) (calibrated for Makibox hot bed)
// 13 is 100k Hisens 3950  1% up to 300°C for hotend "Simple ONE " & "Hotend "All In ONE"
// 20 is the PT100 circuit found in the Ultimainboard V2.x
// 60 is 100k Maker's Tool Works Kapton Bed Thermistor beta=3950
//
//    1k ohm pullup tables - This is not normal, you would have to have changed out your 4.7k for 1k
//                          (but gives greater accuracy and more stable PID)
// 51 is 100k thermistor - EPCOS (1k pullup)
// 52 is 200k thermistor - ATC Semitec 204GT-2 (1k pullup)
// 55 is 100k thermistor - ATC Semitec 104GT-2 (Used in ParCan & J-Head) (1k pullup)
//
// 1047 is Pt1000 with 4k7 pullup
// 1010 is Pt1000 with 1k pullup (non standard)
// 147 is Pt100 with 4k7 pullup
// 110 is Pt100 with 1k pullup (non standard)
// 998 and 999 are Dummy Tables. They will ALWAYS read 25°C or the temperature defined below.
//     Use it for Testing or Development purposes. NEVER for production machine.
//#define DUMMY_THERMISTOR_998_VALUE 25
//#define DUMMY_THERMISTOR_999_VALUE 100
// :{ '0': "Not used", '4': "10k !! do not use for a hotend. Bad resolution at high temp. !!", '1': "100k / 4.7k - EPCOS", '51': "100k / 1k - EPCOS", '6': "100k / 4.7k EPCOS - Not as accurate as Table 1", '5': "100K / 4.7k - ATC Semitec 104GT-2 (Used in ParCan & J-Head)", '7': "100k / 4.7k Honeywell 135-104LAG-J01", '71': "100k / 4.7k Honeywell 135-104LAF-J01", '8': "100k / 4.7k 0603 SMD Vishay NTCS0603E3104FXT", '9': "100k / 4.7k GE Sensing AL03006-58.2K-97-G1", '10': "100k / 4.7k RS 198-961", '11': "100k / 4.7k beta 3950 1%", '12': "100k / 4.7k 0603 SMD Vishay NTCS0603E3104FXT (calibrated for Makibox hot bed)", '13': "100k Hisens 3950  1% up to 300°C for hotend 'Simple ONE ' & hotend 'All In ONE'", '60': "100k Maker's Tool Works Kapton Bed Thermistor beta=3950", '55': "100k / 1k - ATC Semitec 104GT-2 (Used in ParCan & J-Head)", '2': "200k / 4.7k - ATC Semitec 204GT-2", '52': "200k / 1k - ATC Semitec 204GT-2", '-2': "Thermocouple + MAX6675 (only for sensor 0)", '-1': "Thermocouple + AD595", '3': "Mendel-parts / 4.7k", '1047': "Pt1000 / 4.7k", '1010': "Pt1000 / 1k (non standard)", '20': "PT100 (Ultimainboard V2.x)", '147': "Pt100 / 4.7k", '110': "Pt100 / 1k (non-standard)", '998': "Dummy 1", '999': "Dummy 2" }
#define TEMP_SENSOR_0 5
#define TEMP_SENSOR_1 0
#define TEMP_SENSOR_2 0
#define TEMP_SENSOR_3 0
#define TEMP_SENSOR_BED 1

// This makes temp sensor 1 a redundant sensor for sensor 0. If the temperatures difference between these sensors is to high the print will be aborted.
//#define TEMP_SENSOR_1_AS_REDUNDANT
#define MAX_REDUNDANT_TEMP_SENSOR_DIFF 10

// Actual temperature must be close to target for this long before M109 returns success
#define TEMP_RESIDENCY_TIME 10  // (seconds)
#define TEMP_HYSTERESIS 3       // (degC) range of +/- temperatures considered "close" to the target one
#define TEMP_WINDOW     1       // (degC) Window around target to start the residency timer x degC early.

// The minimal temperature defines the temperature below which the heater will not be enabled It is used
// to check that the wiring to the thermistor is not broken.
// Otherwise this would lead to the heater being powered on all the time.
#define HEATER_0_MINTEMP 5
#define HEATER_1_MINTEMP 5
#define HEATER_2_MINTEMP 5
#define HEATER_3_MINTEMP 5
#define BED_MINTEMP 5

// When temperature exceeds max temp, your heater will be switched off.
// This feature exists to protect your hotend from overheating accidentally, but *NOT* from thermistor short/failure!
// You should use MINTEMP for thermistor short/failure protection.
#define HEATER_0_MAXTEMP 290
#define HEATER_1_MAXTEMP 275
#define HEATER_2_MAXTEMP 275
#define HEATER_3_MAXTEMP 275
#define BED_MAXTEMP 130

// If your bed has low resistance e.g. .6 ohm and throws the fuse you can duty cycle it to reduce the
// average current. The value should be an integer and the heat bed will be turned on for 1 interval of
// HEATER_BED_DUTY_CYCLE_DIVIDER intervals.
//#define HEATER_BED_DUTY_CYCLE_DIVIDER 4

// If you want the M105 heater power reported in watts, define the BED_WATTS, and (shared for all extruders) EXTRUDER_WATTS
//#define EXTRUDER_WATTS (12.0*12.0/6.7) //  P=I^2/R
//#define BED_WATTS (12.0*12.0/1.1)      // P=I^2/R

//===========================================================================
//============================= PID Settings ================================
//===========================================================================
// PID Tuning Guide here: http://reprap.org/wiki/PID_Tuning

// Comment the following line to disable PID and enable bang-bang.
#define PIDTEMP
#define BANG_MAX 255 // limits current to nozzle while in bang-bang mode; 255=full current
#define PID_MAX BANG_MAX // limits current to nozzle while PID is active (see PID_FUNCTIONAL_RANGE below); 255=full current
#if ENABLED(PIDTEMP)
  //#define PID_DEBUG // Sends debug data to the serial port.
  //#define PID_OPENLOOP 1 // Puts PID in open loop. M104/M140 sets the output power from 0 to PID_MAX
  //#define SLOW_PWM_HEATERS // PWM with very low frequency (roughly 0.125Hz=8s) and minimum state time of approximately 1s useful for heaters driven by a relay
  //#define PID_PARAMS_PER_EXTRUDER // Uses separate PID parameters for each extruder (useful for mismatched extruders)
                                    // Set/get with gcode: M301 E[extruder number, 0-2]
  #define PID_FUNCTIONAL_RANGE 10 // If the temperature difference between the target temperature and the actual temperature
                                  // is more then PID_FUNCTIONAL_RANGE then the PID will be shut off and the heater will be set to min/max.
  #define PID_INTEGRAL_DRIVE_MAX PID_MAX  //limit for the integral term
  #define K1 0.95 //smoothing factor within the PID

  // If you are using a pre-configured hotend then you can use one of the value sets by uncommenting it
  // Ultimaker
#define  DEFAULT_Kp 30.09
#define  DEFAULT_Ki 2.92
#define  DEFAULT_Kd 77.57

  // MakerGear
  //#define  DEFAULT_Kp 7.0
  //#define  DEFAULT_Ki 0.1
  //#define  DEFAULT_Kd 12

  // Mendel Parts V9 on 12V
  //#define  DEFAULT_Kp 63.0
  //#define  DEFAULT_Ki 2.25
  //#define  DEFAULT_Kd 440

#endif // PIDTEMP

//===========================================================================
//============================= PID > Bed Temperature Control ===============
//===========================================================================
// Select PID or bang-bang with PIDTEMPBED. If bang-bang, BED_LIMIT_SWITCHING will enable hysteresis
//
// Uncomment this to enable PID on the bed. It uses the same frequency PWM as the extruder.
// If your PID_dT is the default, and correct for your hardware/configuration, that means 7.689Hz,
// which is fine for driving a square wave into a resistive load and does not significantly impact you FET heating.
// This also works fine on a Fotek SSR-10DA Solid State Relay into a 250W heater.
// If your configuration is significantly different than this and you don't understand the issues involved, you probably
// shouldn't use bed PID until someone else verifies your hardware works.
// If this is enabled, find your own PID constants below.
#define PIDTEMPBED

//#define BED_LIMIT_SWITCHING

// This sets the max power delivered to the bed, and replaces the HEATER_BED_DUTY_CYCLE_DIVIDER option.
// all forms of bed control obey this (PID, bang-bang, bang-bang with hysteresis)
// setting this to anything other than 255 enables a form of PWM to the bed just like HEATER_BED_DUTY_CYCLE_DIVIDER did,
// so you shouldn't use it unless you are OK with PWM on your bed.  (see the comment on enabling PIDTEMPBED)
#define MAX_BED_POWER 255 // limits duty cycle to bed; 255=full current

//#define PID_BED_DEBUG // Sends debug data to the serial port.

#if ENABLED(PIDTEMPBED)

  #define PID_BED_INTEGRAL_DRIVE_MAX MAX_BED_POWER //limit for the integral term

  //120v 250W silicone heater into 4mm borosilicate (MendelMax 1.5+)
  //from FOPDT model - kp=.39 Tp=405 Tdead=66, Tc set to 79.2, aggressive factor of .15 (vs .1, 1, 10)


//set 1 m304 p45.06 i1.94 d261.37
#define  DEFAULT_bedKp 45.06
#define  DEFAULT_bedKi 1.94
#define  DEFAULT_bedKd 261.37

//set 2 m304 p73.42 i3.09 d436.65
// #define  DEFAULT_bedKp 73.42
// #define  DEFAULT_bedKi 3.09
// #define  DEFAULT_bedKd 436.65

//set 3 m304 p85.57 i9.36 d195.59
// #define  DEFAULT_bedKp 85.57
// #define  DEFAULT_bedKi 9.36
// #define  DEFAULT_bedKd 195.59

//set 4 m304 p26.65 i0.96 d185.36
// #define  DEFAULT_bedKp 26.65
// #define  DEFAULT_bedKi 0.96
// #define  DEFAULT_bedKd 185.36

  //120v 250W silicone heater into 4mm borosilicate (MendelMax 1.5+)
  //from pidautotune
  //#define  DEFAULT_bedKp 97.1
  //#define  DEFAULT_bedKi 1.41
  //#define  DEFAULT_bedKd 1675.16

  // FIND YOUR OWN: "M303 E-1 C8 S90" to run autotune on the bed at 90 degreesC for 8 cycles.
#endif // PIDTEMPBED

// @section extruder

//this prevents dangerous Extruder moves, i.e. if the temperature is under the limit
//can be software-disabled for whatever purposes by
#define PREVENT_DANGEROUS_EXTRUDE
//if PREVENT_DANGEROUS_EXTRUDE is on, you can still disable (uncomment) very long bits of extrusion separately.
#define PREVENT_LENGTHY_EXTRUDE

#define EXTRUDE_MINTEMP 170
#define EXTRUDE_MAXLENGTH (X_MAX_LENGTH+Y_MAX_LENGTH) //prevent extrusion of very large distances.

//===========================================================================
//======================== Thermal Runaway Protection =======================
//===========================================================================

/**
 * Thermal Runaway Protection protects your printer from damage and fire if a
 * thermistor falls out or temperature sensors fail in any way.
 *
 * The issue: If a thermistor falls out or a temperature sensor fails,
 * Marlin can no longer sense the actual temperature. Since a disconnected
 * thermistor reads as a low temperature, the firmware will keep the heater on.
 *
 * The solution: Once the temperature reaches the target, start observing.
 * If the temperature stays too far below the target (hysteresis) for too long,
 * the firmware will halt as a safety precaution.
 */

//#define THERMAL_PROTECTION_HOTENDS // Enable thermal protection for all extruders
//#define THERMAL_PROTECTION_BED     // Enable thermal protection for the heated bed

//===========================================================================
//============================= Mechanical Settings =========================
//===========================================================================

// @section machine

// Uncomment this option to enable CoreXY kinematics
//#define COREXY

// Uncomment this option to enable CoreXZ kinematics
//#define COREXZ

// Enable this option for Toshiba steppers
//#define CONFIG_STEPPERS_TOSHIBA

// @section homing

// coarse Endstop Settings
#define ENDSTOPPULLUPS // Comment this out (using // at the start of the line) to disable the endstop pullup resistors

#if DISABLED(ENDSTOPPULLUPS)
  // fine endstop settings: Individual pullups. will be ignored if ENDSTOPPULLUPS is defined
  //#define ENDSTOPPULLUP_XMAX
  //#define ENDSTOPPULLUP_YMAX
  //#define ENDSTOPPULLUP_ZMAX
  //#define ENDSTOPPULLUP_XMIN
  //#define ENDSTOPPULLUP_YMIN
  //#define ENDSTOPPULLUP_ZMIN
  //#define ENDSTOPPULLUP_ZMIN_PROBE
#endif

// Mechanical endstop with COM to ground and NC to Signal uses "false" here (most common setup).
const bool X_MIN_ENDSTOP_INVERTING = true; // set to true to invert the logic of the endstop.
const bool Y_MIN_ENDSTOP_INVERTING = true; // set to true to invert the logic of the endstop.
const bool Z_MIN_ENDSTOP_INVERTING = true; // set to true to invert the logic of the endstop.
const bool X_MAX_ENDSTOP_INVERTING = true; // set to true to invert the logic of the endstop.
const bool Y_MAX_ENDSTOP_INVERTING = true; // set to true to invert the logic of the endstop.
const bool Z_MAX_ENDSTOP_INVERTING = true; // set to true to invert the logic of the endstop.
const bool Z_MIN_PROBE_ENDSTOP_INVERTING = false; // set to true to invert the logic of the endstop.
//#define DISABLE_MAX_ENDSTOPS
//#define DISABLE_MIN_ENDSTOPS

// If you want to enable the Z probe pin, but disable its use, uncomment the line below.
// This only affects a Z probe endstop if you have separate Z min endstop as well and have
// activated Z_MIN_PROBE_ENDSTOP below. If you are using the Z Min endstop on your Z probe,
// this has no effect.
//#define DISABLE_Z_MIN_PROBE_ENDSTOP

// For Inverting Stepper Enable Pins (Active Low) use 0, Non Inverting (Active High) use 1
// :{0:'Low',1:'High'}
#define X_ENABLE_ON 0
#define Y_ENABLE_ON 0
#define Z_ENABLE_ON 0
#define E_ENABLE_ON 0 // For all extruders

// Disables axis when it's not being used.
// WARNING: When motors turn off there is a chance of losing position accuracy!
#define DISABLE_X false
#define DISABLE_Y false
#define DISABLE_Z false

// @section extruder

#define DISABLE_E false // For all extruders
#define DISABLE_INACTIVE_EXTRUDER true //disable only inactive extruders and keep active extruder enabled

// @section machine

// Invert the stepper direction. Change (or reverse the motor connector) if an axis goes the wrong way.
#define INVERT_X_DIR true
#define INVERT_Y_DIR false
#define INVERT_Z_DIR true

// @section extruder

// For direct drive extruder v9 set to true, for geared extruder set to false.
#define INVERT_E0_DIR false
#define INVERT_E1_DIR false
#define INVERT_E2_DIR false
#define INVERT_E3_DIR false

// @section homing

// ENDSTOP SETTINGS:
// Sets direction of endstops when homing; 1=MAX, -1=MIN
// :[-1,1]
#define X_HOME_DIR -1
#define Y_HOME_DIR -1
#define Z_HOME_DIR -1

#define min_software_endstops true // If true, axis won't move to coordinates less than HOME_POS.
#define max_software_endstops true  // If true, axis won't move to coordinates greater than the defined lengths below.

// @section machine

// Travel limits after homing (units are in mm)
#define X_MIN_POS 0
#define Y_MIN_POS 0
#define Z_MIN_POS 0
#define X_MAX_POS 285
#define Y_MAX_POS 218
#define Z_MAX_POS 300

//===========================================================================
//========================= Filament Runout Sensor ==========================
//===========================================================================
//#define FILAMENT_RUNOUT_SENSOR // Uncomment for defining a filament runout sensor such as a mechanical or opto endstop to check the existence of filament
                                 // In RAMPS uses servo pin 2. Can be changed in pins file. For other boards pin definition should be made.
                                 // It is assumed that when logic high = filament available
                                 //                    when logic  low = filament ran out
#if ENABLED(FILAMENT_RUNOUT_SENSOR)
  const bool FIL_RUNOUT_INVERTING = true;  // Should be uncommented and true or false should assigned
  #define ENDSTOPPULLUP_FIL_RUNOUT // Uncomment to use internal pullup for filament runout pins if the sensor is defined.
  #define FILAMENT_RUNOUT_SCRIPT "M600"
#endif

//===========================================================================
//=========================== Manual Bed Leveling ===========================
//===========================================================================

//#define MANUAL_BED_LEVELING  // Add display menu option for bed leveling.
//#define MESH_BED_LEVELING    // Enable mesh bed leveling.

#if ENABLED(MANUAL_BED_LEVELING)
  #define MBL_Z_STEP 0.025  // Step size while manually probing Z axis.
#endif  // MANUAL_BED_LEVELING

#if ENABLED(MESH_BED_LEVELING)
  #define MESH_MIN_X 10
  #define MESH_MAX_X (X_MAX_POS - MESH_MIN_X)
  #define MESH_MIN_Y 10
  #define MESH_MAX_Y (Y_MAX_POS - MESH_MIN_Y)
  #define MESH_NUM_X_POINTS 3  // Don't use more than 7 points per axis, implementation limited.
  #define MESH_NUM_Y_POINTS 3
  #define MESH_HOME_SEARCH_Z 4  // Z after Home, bed somewhere below but above 0.0.
#endif  // MESH_BED_LEVELING

//===========================================================================
//============================ Bed Auto Leveling ============================
//===========================================================================

// @section bedlevel

#define AUTO_BED_LEVELING_FEATURE // Delete the comment to enable (remove // at the start of the line)
#define DEBUG_LEVELING_FEATURE
#define Z_MIN_PROBE_REPEATABILITY_TEST  // If not commented out, Z-Probe Repeatability test will be included if Auto Bed Leveling is Enabled.

#if ENABLED(AUTO_BED_LEVELING_FEATURE)

  // There are 2 different ways to specify probing locations:
  //
  // - "grid" mode
  //   Probe several points in a rectangular grid.
  //   You specify the rectangle and the density of sample points.
  //   This mode is preferred because there are more measurements.
  //
  // - "3-point" mode
  //   Probe 3 arbitrary points on the bed (that aren't colinear)
  //   You specify the XY coordinates of all 3 points.

  // Enable this to sample the bed in a grid (least squares solution).
  // Note: this feature generates 10KB extra code size.
  #define AUTO_BED_LEVELING_GRID

  #if ENABLED(AUTO_BED_LEVELING_GRID)

    #define LEFT_PROBE_BED_POSITION 15
    #define RIGHT_PROBE_BED_POSITION 185
    #define FRONT_PROBE_BED_POSITION 20
    #define BACK_PROBE_BED_POSITION 175

    #define MIN_PROBE_EDGE 10 // The Z probe minimum square sides can be no smaller than this.

    // Set the number of grid points per dimension.
    // You probably don't need more than 3 (squared=9).
    #define AUTO_BED_LEVELING_GRID_POINTS 3

  #else  // !AUTO_BED_LEVELING_GRID

      // Arbitrary points to probe.
      // A simple cross-product is used to estimate the plane of the bed.
      #define ABL_PROBE_PT_1_X 15
      #define ABL_PROBE_PT_1_Y 180
      #define ABL_PROBE_PT_2_X 15
      #define ABL_PROBE_PT_2_Y 20
      #define ABL_PROBE_PT_3_X 170
      #define ABL_PROBE_PT_3_Y 20

  #endif // AUTO_BED_LEVELING_GRID

  // Offsets to the Z probe relative to the nozzle tip.
  // X and Y offsets must be integers.
  #define X_PROBE_OFFSET_FROM_EXTRUDER -46     // Z probe to nozzle X offset: -left  +right
  #define Y_PROBE_OFFSET_FROM_EXTRUDER -9     // Z probe to nozzle Y offset: -front +behind
  #define Z_PROBE_OFFSET_FROM_EXTRUDER -2.85  // Z probe to nozzle Z offset: -below (always!)

  #define Z_RAISE_BEFORE_HOMING 5       // (in mm) Raise Z axis before homing (G28) for Z probe clearance.
                                        // Be sure you have this distance over your Z_MAX_POS in case.

  #define XY_TRAVEL_SPEED 8000         // X and Y axis travel speed between probes, in mm/min.

  #define Z_RAISE_BEFORE_PROBING 10   // How much the Z axis will be raised before traveling to the first probing point.
  #define Z_RAISE_BETWEEN_PROBINGS 5  // How much the Z axis will be raised when traveling from between next probing points.
  #define Z_RAISE_AFTER_PROBING 10    // How much the Z axis will be raised after the last probing point.

//#define Z_PROBE_END_SCRIPT "G1 Z10 F12000\nG1 X15 Y330\nG1 Z0.5\nG1 Z10" // These commands will be executed in the end of G29 routine.
                                                                            // Useful to retract a deployable Z probe.

  //#define Z_PROBE_SLED // Turn on if you have a Z probe mounted on a sled like those designed by Charles Bell.
  //#define SLED_DOCKING_OFFSET 5 // The extra distance the X axis must travel to pickup the sled. 0 should be fine but you can push it further if you'd like.


  //If you have enabled the Bed Auto Leveling and are using the same Z Probe for Z Homing,
  //it is highly recommended you let this Z_SAFE_HOMING enabled!!!

  #define Z_SAFE_HOMING   // This feature is meant to avoid Z homing with Z probe outside the bed area.
                          // When defined, it will:
                          // - Allow Z homing only after X and Y homing AND stepper drivers still enabled.
                          // - If stepper drivers timeout, it will need X and Y homing again before Z homing.
                          // - Position the Z probe in a defined XY point before Z Homing when homing all axis (G28).
                          // - Block Z homing only when the Z probe is outside bed area.

  #if ENABLED(Z_SAFE_HOMING)

    #define Z_SAFE_HOMING_X_POINT ((X_MIN_POS + X_MAX_POS) / 2)    // X point for Z homing when homing all axis (G28).
    #define Z_SAFE_HOMING_Y_POINT ((Y_MIN_POS + Y_MAX_POS) / 2)    // Y point for Z homing when homing all axis (G28).

  #endif

  // Support for a dedicated Z probe endstop separate from the Z min endstop.
  // If you would like to use both a Z probe and a Z min endstop together,
  // uncomment #define Z_MIN_PROBE_ENDSTOP and read the instructions below.
  // If you still want to use the Z min endstop for homing, disable Z_SAFE_HOMING above.
  // Example: To park the head outside the bed area when homing with G28.
  //
  // WARNING:
  // The Z min endstop will need to set properly as it would without a Z probe
  // to prevent head crashes and premature stopping during a print.
  //
  // To use a separate Z probe endstop, you must have a Z_MIN_PROBE_PIN
  // defined in the pins_XXXXX.h file for your control board.
  // If you are using a servo based Z probe, you will need to enable NUM_SERVOS,
  // Z_ENDSTOP_SERVO_NR and SERVO_ENDSTOP_ANGLES in the R/C SERVO support below.
  // RAMPS 1.3/1.4 boards may be able to use the 5V, Ground and the D32 pin
  // in the Aux 4 section of the RAMPS board. Use 5V for powered sensors,
  // otherwise connect to ground and D32 for normally closed configuration
  // and 5V and D32 for normally open configurations.
  // Normally closed configuration is advised and assumed.
  // The D32 pin in Aux 4 on RAMPS maps to the Arduino D32 pin.
  // Z_MIN_PROBE_PIN is setting the pin to use on the Arduino.
  // Since the D32 pin on the RAMPS maps to D32 on Arduino, this works.
  // D32 is currently selected in the RAMPS 1.3/1.4 pin file.
  // All other boards will need changes to the respective pins_XXXXX.h file.
  //
  // WARNING:
  // Setting the wrong pin may have unexpected and potentially disastrous outcomes.
  // Use with caution and do your homework.
  //
  //#define Z_MIN_PROBE_ENDSTOP

#endif // AUTO_BED_LEVELING_FEATURE


// @section homing

// The position of the homing switches
//#define MANUAL_HOME_POSITIONS  // If defined, MANUAL_*_HOME_POS below will be used
//#define BED_CENTER_AT_0_0  // If defined, the center of the bed is at (X=0, Y=0)

// Manual homing switch locations:
// For deltabots this means top and center of the Cartesian print volume.
#if ENABLED(MANUAL_HOME_POSITIONS)
  #define MANUAL_X_HOME_POS 0
  #define MANUAL_Y_HOME_POS 0
  #define MANUAL_Z_HOME_POS 0
  //#define MANUAL_Z_HOME_POS 402 // For delta: Distance between nozzle and print surface after homing.
#endif

// @section movement

/**
 * MOVEMENT SETTINGS
 */

#define HOMING_FEEDRATE {50*45, 50*45, 4*45, 0}  // set the homing speeds (mm/min)

// default settings


#define DEFAULT_AXIS_STEPS_PER_UNIT   {157.48, 157.48, 812, 95}  // steps/mm using 1/32 microstepping aka drv8825
#define DEFAULT_MAX_FEEDRATE          {400, 400, 4, 45}    // (mm/sec)
#define DEFAULT_MAX_ACCELERATION      {5000,5000,50,5000}    // X, Y, Z, E maximum start speed for accelerated moves. E default values are good for Skeinforge 40+, for older versions raise them a lot.

#define DEFAULT_ACCELERATION          1000    // X, Y, Z and E acceleration in mm/s^2 for printing moves
#define DEFAULT_RETRACT_ACCELERATION  2000    // E acceleration in mm/s^2 for retracts
#define DEFAULT_TRAVEL_ACCELERATION   3000    // X, Y, Z acceleration in mm/s^2 for travel (non printing) moves

// The speed change that does not require acceleration (i.e. the software might assume it can be done instantaneously)
#define DEFAULT_XYJERK                15.0    // (mm/sec)
#define DEFAULT_ZJERK                 0.4     // (mm/sec)
#define DEFAULT_EJERK                 5.0    // (mm/sec)


//=============================================================================
//============================= Additional Features ===========================
//=============================================================================

// @section more

// Custom M code points
#define CUSTOM_M_CODES
#if ENABLED(CUSTOM_M_CODES)
  #if ENABLED(AUTO_BED_LEVELING_FEATURE)
    #define CUSTOM_M_CODE_SET_Z_PROBE_OFFSET 851
    #define Z_PROBE_OFFSET_RANGE_MIN -20
    #define Z_PROBE_OFFSET_RANGE_MAX 20
  #endif
#endif

// @section extras

// EEPROM
// The microcontroller can store settings in the EEPROM, e.g. max velocity...
// M500 - stores parameters in EEPROM
// M501 - reads parameters from EEPROM (if you need reset them after you changed them temporarily).
// M502 - reverts to the default "factory settings".  You still need to store them in EEPROM afterwards if you want to.
//define this to enable EEPROM support
#define EEPROM_SETTINGS

#if ENABLED(EEPROM_SETTINGS)
  // To disable EEPROM Serial responses and decrease program space by ~1700 byte: comment this out:
  #define EEPROM_CHITCHAT // Please keep turned on if you can.
#endif

//
// M100 Free Memory Watcher
//
//#define M100_FREE_MEMORY_WATCHER // uncomment to add the M100 Free Memory Watcher for debug purpose

// @section temperature

// Preheat Constants
#define PLA_PREHEAT_HOTEND_TEMP 180
#define PLA_PREHEAT_HPB_TEMP 70
#define PLA_PREHEAT_FAN_SPEED 0   // Insert Value between 0 and 255

#define ABS_PREHEAT_HOTEND_TEMP 240
#define ABS_PREHEAT_HPB_TEMP 110
#define ABS_PREHEAT_FAN_SPEED 0   // Insert Value between 0 and 255

//==============================LCD and SD support=============================
// @section lcd

// Define your display language below. Replace (en) with your language code and uncomment.
// en, pl, fr, de, es, ru, bg, it, pt, pt-br, fi, an, nl, ca, eu, kana, kana_utf8, cn, test
// See also language.h
#define LANGUAGE_INCLUDE GENERATE_LANGUAGE_INCLUDE(en)

// Choose ONE of these 3 charsets. This has to match your hardware. Ignored for full graphic display.
// To find out what type you have - compile with (test) - upload - click to get the menu. You'll see two typical lines from the upper half of the charset.
// See also documentation/LCDLanguageFont.md
  #define DISPLAY_CHARSET_HD44780_JAPAN        // this is the most common hardware
  //#define DISPLAY_CHARSET_HD44780_WESTERN
  //#define DISPLAY_CHARSET_HD44780_CYRILLIC

//#define ULTRA_LCD  //general LCD support, also 16x2
//#define DOGLCD  // Support for SPI LCD 128x64 (Controller ST7565R graphic Display Family)
//#define SDSUPPORT // Enable SD Card Support in Hardware Console
// Changed behaviour! If you need SDSUPPORT uncomment it!
//#define SDSLOW // Use slower SD transfer mode (not normally needed - uncomment if you're getting volume init error)
//#define SDEXTRASLOW // Use even slower SD transfer mode (not normally needed - uncomment if you're getting volume init error)
//#define SD_CHECK_AND_RETRY // Use CRC checks and retries on the SD communication
//#define ENCODER_PULSES_PER_STEP 1 // Increase if you have a high resolution encoder
//#define ENCODER_STEPS_PER_MENU_ITEM 5 // Set according to ENCODER_PULSES_PER_STEP or your liking
//#define ULTIMAKERCONTROLLER //as available from the Ultimaker online store.
//#define ULTIPANEL  //the UltiPanel as on Thingiverse
//#define SPEAKER // The sound device is a speaker - not a buzzer. A buzzer resonates with his own frequency.
//#define LCD_FEEDBACK_FREQUENCY_DURATION_MS 100 // the duration the buzzer plays the UI feedback sound. ie Screen Click
//#define LCD_FEEDBACK_FREQUENCY_HZ 1000         // this is the tone frequency the buzzer plays when on UI feedback. ie Screen Click
                                                 // 0 to disable buzzer feedback. Test with M300 S<frequency Hz> P<duration ms>
// PanelOne from T3P3 (via RAMPS 1.4 AUX2/AUX3)
// http://reprap.org/wiki/PanelOne
//#define PANEL_ONE

// The MaKr3d Makr-Panel with graphic controller and SD support
// http://reprap.org/wiki/MaKr3d_MaKrPanel
//#define MAKRPANEL

// The Panucatt Devices Viki 2.0 and mini Viki with Graphic LCD
// http://panucatt.com
// ==> REMEMBER TO INSTALL U8glib to your ARDUINO library folder: http://code.google.com/p/u8glib/wiki/u8glib
//#define VIKI2
//#define miniVIKI

// This is a new controller currently under development.  https://github.com/eboston/Adafruit-ST7565-Full-Graphic-Controller/
//
// ==> REMEMBER TO INSTALL U8glib to your ARDUINO library folder: http://code.google.com/p/u8glib/wiki/u8glib
//#define ELB_FULL_GRAPHIC_CONTROLLER
//#define SD_DETECT_INVERTED

// The RepRapDiscount Smart Controller (white PCB)
// http://reprap.org/wiki/RepRapDiscount_Smart_Controller
#define REPRAP_DISCOUNT_SMART_CONTROLLER

// The GADGETS3D G3D LCD/SD Controller (blue PCB)
// http://reprap.org/wiki/RAMPS_1.3/1.4_GADGETS3D_Shield_with_Panel
//#define G3D_PANEL

// The RepRapDiscount FULL GRAPHIC Smart Controller (quadratic white PCB)
// http://reprap.org/wiki/RepRapDiscount_Full_Graphic_Smart_Controller
//
// ==> REMEMBER TO INSTALL U8glib to your ARDUINO library folder: http://code.google.com/p/u8glib/wiki/u8glib
//#define REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER

// The RepRapWorld REPRAPWORLD_KEYPAD v1.1
// http://reprapworld.com/?products_details&products_id=202&cPath=1591_1626
//#define REPRAPWORLD_KEYPAD
//#define REPRAPWORLD_KEYPAD_MOVE_STEP 10.0 // how much should be moved when a key is pressed, eg 10.0 means 10mm per click

// The Elefu RA Board Control Panel
// http://www.elefu.com/index.php?route=product/product&product_id=53
// REMEMBER TO INSTALL LiquidCrystal_I2C.h in your ARDUINO library folder: https://github.com/kiyoshigawa/LiquidCrystal_I2C
//#define RA_CONTROL_PANEL

// The MakerLab Mini Panel with graphic controller and SD support
// http://reprap.org/wiki/Mini_panel
//#define MINIPANEL

/**
 * I2C Panels
 */

//#define LCD_I2C_SAINSMART_YWROBOT

// PANELOLU2 LCD with status LEDs, separate encoder and click inputs
//
// This uses the LiquidTWI2 library v1.2.3 or later ( https://github.com/lincomatic/LiquidTWI2 )
// Make sure the LiquidTWI2 directory is placed in the Arduino or Sketchbook libraries subdirectory.
// (v1.2.3 no longer requires you to define PANELOLU in the LiquidTWI2.h library header file)
// Note: The PANELOLU2 encoder click input can either be directly connected to a pin
//       (if BTN_ENC defined to != -1) or read through I2C (when BTN_ENC == -1).
//#define LCD_I2C_PANELOLU2

// Panucatt VIKI LCD with status LEDs, integrated click & L/R/U/P buttons, separate encoder inputs
//#define LCD_I2C_VIKI

// SSD1306 OLED generic display support
// ==> REMEMBER TO INSTALL U8glib to your ARDUINO library folder: http://code.google.com/p/u8glib/wiki/u8glib
//#define U8GLIB_SSD1306

// Shift register panels
// ---------------------
// 2 wire Non-latching LCD SR from:
// https://bitbucket.org/fmalpartida/new-liquidcrystal/wiki/schematics#!shiftregister-connection
// LCD configuration: http://reprap.org/wiki/SAV_3D_LCD
//#define SAV_3DLCD

// @section extras

// Increase the FAN pwm frequency. Removes the PWM noise but increases heating in the FET/Arduino
//#define FAST_PWM_FAN

// Use software PWM to drive the fan, as for the heaters. This uses a very low frequency
// which is not as annoying as with the hardware PWM. On the other hand, if this frequency
// is too low, you should also increment SOFT_PWM_SCALE.
//#define FAN_SOFT_PWM

// Incrementing this by 1 will double the software PWM frequency,
// affecting heaters, and the fan if FAN_SOFT_PWM is enabled.
// However, control resolution will be halved for each increment;
// at zero value, there are 128 effective control positions.
#define SOFT_PWM_SCALE 0

// Temperature status LEDs that display the hotend and bet temperature.
// If all hotends and bed temperature and temperature setpoint are < 54C then the BLUE led is on.
// Otherwise the RED led is on. There is 1C hysteresis.
//#define TEMP_STAT_LEDS

// M240  Triggers a camera by emulating a Canon RC-1 Remote
// Data from: http://www.doc-diy.net/photo/rc-1_hacked/
//#define PHOTOGRAPH_PIN     23

// SkeinForge sends the wrong arc g-codes when using Arc Point as fillet procedure
//#define SF_ARC_FIX

// Support for the BariCUDA Paste Extruder.
//#define BARICUDA

//define BlinkM/CyzRgb Support
//#define BLINKM

/*********************************************************************\
* R/C SERVO support
* Sponsored by TrinityLabs, Reworked by codexmas
**********************************************************************/

// Number of servos
//
// If you select a configuration below, this will receive a default value and does not need to be set manually
// set it manually if you have more servos than extruders and wish to manually control some
// leaving it undefined or defining as 0 will disable the servo subsystem
// If unsure, leave commented / disabled
//
//#define NUM_SERVOS 4// Servo index starts with 0 for M280 command

// Servo Endstops
//
// This allows for servo actuated endstops, primary usage is for the Z Axis to eliminate calibration or bed height changes.
// Use M851 to set the Z probe vertical offset from the nozzle. Store that setting with M500.
//
#define X_ENDSTOP_SERVO_NR 1
#define Y_ENDSTOP_SERVO_NR 2
#define Z_ENDSTOP_SERVO_NR 0
#define SERVO_ENDSTOP_ANGLES {{0,0}, {0,0}, {9,90}} // X,Y,Z Axis Extend and Retract angles

// Servo deactivation
//
// With this option servos are powered only during movement, then turned off to prevent jitter.
//#define DEACTIVATE_SERVOS_AFTER_MOVE

#if ENABLED(DEACTIVATE_SERVOS_AFTER_MOVE)
  // Delay (in microseconds) before turning the servo off. This depends on the servo speed.
  // 300ms is a good value but you can try less delay.
  // If the servo can't reach the requested position, increase it.
  #define SERVO_DEACTIVATION_DELAY 1250
#endif

/**********************************************************************\
 * Support for a filament diameter sensor
 * Also allows adjustment of diameter at print time (vs  at slicing)
 * Single extruder only at this point (extruder 0)
 *
 * Motherboards
 * 34 - RAMPS1.4 - uses Analog input 5 on the AUX2 connector
 * 81 - Printrboard - Uses Analog input 2 on the Exp1 connector (version B,C,D,E)
 * 301 - Rambo  - uses Analog input 3
 * Note may require analog pins to be defined for different motherboards
 **********************************************************************/
// Uncomment below to enable
//#define FILAMENT_SENSOR

#define FILAMENT_SENSOR_EXTRUDER_NUM 0   //The number of the extruder that has the filament sensor (0,1,2)
#define MEASUREMENT_DELAY_CM        14   //measurement delay in cm.  This is the distance from filament sensor to middle of barrel

#define DEFAULT_NOMINAL_FILAMENT_DIA 3.00  //Enter the diameter (in mm) of the filament generally used (3.0 mm or 1.75 mm) - this is then used in the slicer software.  Used for sensor reading validation
#define MEASURED_UPPER_LIMIT         3.30  //upper limit factor used for sensor reading validation in mm
#define MEASURED_LOWER_LIMIT         1.90  //lower limit factor for sensor reading validation in mm
#define MAX_MEASUREMENT_DELAY       20     //delay buffer size in bytes (1 byte = 1cm)- limits maximum measurement delay allowable (must be larger than MEASUREMENT_DELAY_CM  and lower number saves RAM)

//defines used in the code
#define DEFAULT_MEASURED_FILAMENT_DIA  DEFAULT_NOMINAL_FILAMENT_DIA  //set measured to nominal initially

//When using an LCD, uncomment the line below to display the Filament sensor data on the last line instead of status.  Status will appear for 5 sec.
//#define FILAMENT_LCD_DISPLAY

#include "Configuration_adv.h"
#include "thermistortables.h"

#endif //CONFIGURATION_H
```

**Comment by thinkyhead** [comment]
What results do you get from `M48` "Repeatability Test"? First thing is to make sure you have a reliable measurement device. If it's not very reliable, you may be able to slow down the probe and get better results.

**Comment by sarf2k4** [comment]
Here is the m48 reports, I did 5 points. These are when the bed are cold

Front Left
Mean: -9.775862
Standard Deviation: 0.008708

Front Right
Mean: -9.939288
Standard Deviation: 0.005591

Back Left
Mean: -10.120197
Standard Deviation: 0.004413

Back Right
Mean: -10.012562
Standard Deviation: 0.007282

Center
Mean: -9.957020
Standard Deviation: 0.005061

The following results are when the bed at 110'c
Eqn coefficients: a: 0.00106733 b: -0.00224448 d: 7.03367090
planeNormal x: -0.001067 y: 0.002244 z: 1.000000

Bed Level Correction Matrix:
+0.999999 +0.000000 +0.001067
+0.000002 +0.999997 -0.002244
-0.001067 +0.002244 +0.999997

M48 at bed 110c
Front Left
Mean: 1.812931
Standard Deviation: 0.011021

Front Right
Mean: 1.695320
Standard Deviation: 0.011357

Back left
Mean: 1.472167
Standard Deviation: 0.010496

Back Right
Mean: 1.581527
Standard Deviation: 0.007527

Center
Mean: 1.581527
Standard Deviation: 0.007527

**Comment by Roxy-3D** [comment]
Those 'Standard Deviation' numbers look acceptable.  (They are not super good...  But they should be good enough)   You do have a fair amount of tilt to your bed.   The Auto Bed Leveling should be able to correct for that, but it would be wise to get the bed more physically level if you can.

**Comment by thinkyhead** [comment]
Time to test with the latest `RCBugFix`. I suggest more adjustment, testing, adjustment, testing, and adjustment. I don't think we're seeing anything like a software issue here.
